repository/http: add a custom pre-send hook

lucab opened the pull request:
This adds an optional user-provided boxed closure to `HttpRepository`
which can be triggered to customize an HTTP request before sending it.

Closes https://github.com/heartsucker/rust-tuf/issues/99

lucab committed:
repository/http: add a custom pre-send hook

This adds an optional user-provided boxed closure to `HttpRepository`
which can be triggered to customize an HTTP request before sending it.

heartsucker commented on src/repository.rs:
```diff
@@ -317,11 +321,18 @@ where
     }
 
+    /// Set an hook function to be invoked before sending an HTTP request.
+    pub fn set_http_hook(&mut self, hook: Option<HttpHook>) -> &Self {
```
Pedantic: I'm trying to avoid the Java `set_xxx` convention, so could you change this to just `http_hook`.

Less pedantic: Could you put a short example in the docs showing how someone might use this to set auth headers (or whatever your use case was). It possible the description isn't enough.

lucab replied:
* I also don't like getters&setters either, but I thought the non-prefixed one is typically used for getters not for setters. But I will amend it.
 * Sure. Is something similar to the test-case (minus the panic part) ok?

heartsucker replied:
Oops. Just saw this. I'm not using getters since this is a builder and presumably people know what they're putting in before they do, so I won't need to support both.

And yeah, something similar to your test case but without the panic and with an actual manipulation.

heartsucker commented on src/repository.rs:
```diff
@@ -591,4 +606,23 @@ mod test {
     }
+
+    #[test]
+    #[should_panic]
```
I also wonder if this could use a `panic::catch_unwind` just on the final step to ensure we only catch the correct panic.

lucab replied:
Uhm, I never tried this but it seems feasable and reasonable, I'll have a look at it.

lucab commented:
Full disclosure: I didn't test it in my docker usecase yet as I coded while travelling/offline and you forewarned me it probably won't interoperate anyway, but I still wanted to code this and get a review. Do you prefer to wait for a full roundtrip test in my scenario before landing this?

heartsucker commented on src/repository.rs:
```diff
@@ -289,6 +290,8 @@ where
 }
 
+/// A custom hook to be invoked before sending HTTP requests.
+pub type HttpHook = Box<for<'a> Fn(RequestBuilder<'a>) -> RequestBuilder<'a>>;
```
I was thinking about this some more. I think the signatures should be

```rust
// non-public type
type HttpHook = Box<for<'a> Fn(&mut RequestBuilder<'a>) -> Result<RequestBuilder<'a>, Into<Error>>; 
```

And the function docs needs to say that it **must not** panic. I think this is better because I would rather consumers of this lib not introduce panicking code. On top of this, I think it would be good to wrap the call site itself in a `panic::catch_unwind` and returning `Error::Opaque("Hook panicked.".into())` to contain the panic and stop it from spreading elsewhere in the code.

heartsucker replied:
Ok, so I did my research and realized I misunderstood `catch_unwind`. Don't wrap call site, actually. Turns out that's a bad idea.

heartsucker commented:
> Do you prefer to wait for a full roundtrip test in my scenario before landing this?

Yes. It would be good to ensure that it suits your use case before we put this into the lib. Transcribing your use case into the docs would be helpful so that other people see how to use the hook.

lucab commented:
Ack, I will then proceed to address the comment from the review, push it, make sure it can authn to a docker notary and then report back. The full scenario is a bit too large to fit (multi-steps flow to retrieve an authn token), but it can be simplified in a bearer token header (similar to the testcase).

heartsucker commented:
Ok, so maybe not the *whole* case then, but a partial case would be nice. I'm also not 100% sure about what do with my `panic::catch_unwind` proposal as I generally just completely avoid panics in my code, so I'm not sure if what I said is even idiomatic or makes sense. Though I think it does. It would be nice if someone like @brson or a more seasoned Rust dev could give some insight here.

Edit: see other comments. no `catch_unwind`s, plz.

heartsucker commented:
Heya. Thought about this some more. I have the feature in here were I set the lib version in the user agent, and your method would allow someone to overwrite that. This is actually pretty important telemetry for ensuring clients in the field aren't cut off from updates because of this lib making breaking changes.

Also, it should remove the possibility to set the body (since we're only `GET`-ing). I almost think it would be better to give a `Result<Headers, Into<Error>>`. Though someone might need to set query string params as well, so we might need to get back a `Result<(HashMap<String, String>, Headers), Into<Error>>` and now we're getting very complicated.

I'm more hesitant about this PR now that I've run through it a few times.

lucab commented:
Yes, this is a pre-send hook which means that you (as a library author) trust me (as the consumer/service owner) to know better than you about what my environment quirks are, and override some of your decisions. Coming back to your examples, my TUF server may have ~~weird~~over-smart LB in front which requires exact hardcoded user-agent, or some body content in the GET. As unfortunate, extreme and wrong as it sounds, this unfortunately may happen when interacting with somebody else's infra.

Back to a technical point, I think that two possible approaches are:
 * embrace the fact that users may need to override your choices, and let them do it (current PR)
 * try to undo some of the user changes, by invoking the hook earlier and later undoing what you don't like

I find the latter a bit wrong as a library consumer, but if you are mostly concerned with an untouched user-agent I think I can swap the hook order and the clean+set the user-agent. I'd still prefer current approach, though.

> Also, it should remove the possibility to set the body (since we're only GET-ing)

Quoting from memory without RFC at hand, this is not an invariant. GETs *may* have body content and servers *should* ignore it.

> Though someone might need to set query string params as well

This is out of scope from this PR. If somebody needs this, we first have to capture the requirements (is the query dynamic/static? Is it homogeneous across all endpoints/actions?) and then think about how to expose it. I'm not trying to tackling it here, as I wouldn't be able to make the right call.

heartsucker commented:
Ok, those are fair points. I guess as lib author I like to limit the number possible dumb things someone could do. If this PR works for you, then if you address the couple of small issues I pointed out earlier, I'll merge it. I might hide it behind an unstable flag, however.

lucab commented:
Good news: this works, tested with the gist at https://gist.github.com/lucab/73a3519e2086a27cd21cbbf50abed6ff (ignore the fact it is a silly example, I just needed some debug-printf for the final manifest hash).
Bad news: as forewarned I hit https://github.com/heartsucker/rust-tuf/issues/93#issuecomment-312704968, more on this in a separate thread.

For review: second iteration addressing your previous comments, PTAL. Changes:
 * renamed method and tweaked to return a Result type
 * added a proper example based on the above gist with an opaque bearer token
 * wrapped the panic-ing part of the test in its own catch block

lucab commented:
Bump, what should we do with this? I amended as suggested, but it turns out in the end I'm not going to actively consume this or this library. Do you think it is worth anyway to land it for other authenticated TUF endpoints or should I just close it?

heartsucker commented:
Yeah, sorry I haven't been able to look at this. I think it's a good idea, and I think a future version of `rust-tuf` will include it, but I think we will switch to `hyper` `0.11.x` before this is actually needed. I think we can close it for now and figure out what else to do later. If you agree, I'll let you close.

lucab commented:
Ack, closing it. Yes, I hope reqwest at some point will grow direct support for this kind of hooks.
